cmd/arc_summary/arc_summary3.py

```diff
@@ -0,0 +1,833 @@
+               [10**6, 'M'],   # mega  (million)
+               [10**3, 'k'])   # kilo  (thousand)
+
+    hits = int(hits_string)
+
+    if hits >= 1000:
+        for limit, symbol in numbers:
+
+            if hits >= limit:
+                value = hits/limit
+                break
+
+        result = "%0.1f%s" % (value, symbol)
+    else:
+        result = "%d" % hits
+
+    return result
+
+
+def f_perc(value1, value2):
+    """Calculate percentage and return in human-readable form. If
+    rounding produces the result '0.0' though the first number is
+    not zero, include a 'less-than' symbol to avoid confusion.
+    Division by zero is handled by returning 'n/a'; no error
+    is called.
+    """
+
+    v1 = float(value1)
+    v2 = float(value2)
+
+    try:
+        perc = 100 * v1/v2
+    except ZeroDivisionError:
+        result = 'n/a'
+    else:
+        result = '{0:0.1f} %'.format(perc)
+
+    if result == '0.0 %' and v1 > 0:
+        result = '< 0.1 %'
+
+    return result
+
+
+def format_raw_line(name, value):
+    """For the --raw option for the tunable and SPL outputs, decide on the
+    correct formatting based on the --alternate flag.
+    """
+
+    if ARGS.alt:
+        result = '{0}{1}={2}'.format(INDENT, name, value)
+    else:
+        spc = LINE_LENGTH-(len(INDENT)+len(value))
+        result = '{0}{1:<{spc}}{2}'.format(INDENT, name, value, spc=spc)
+
+    return result
+
+
+def get_kstats():
+    """Collect information on the ZFS subsystem from the /proc Linux virtual
+    file system. The step does not perform any further processing, giving us
+    the option to only work on what is actually needed. The name "kstat" is a
+    holdover from the Solaris utility of the same name.
+    """
+
+    result = {}
+    secs = SECTION_PATHS.values()
+
+    for section in secs:
+
+        with open(PROC_PATH+section, 'r') as proc_location:
+            lines = [line for line in proc_location]
+
+        del lines[0:2]  # Get rid of header
+        result[section] = lines
+
+    return result
+
+
+def get_spl_tunables(PATH):
+    """Collect information on the Solaris Porting Layer (SPL) or the
+    tunables, depending on the PATH given. Does not check if PATH is
+    legal.
+    """
+
+    result = {}
+    parameters = os.listdir(PATH)
+
+    for name in parameters:
+
+        with open(PATH+name, 'r') as para_file:
+            value = para_file.read()
+            result[name] = value.strip()
+
+    return result
+
+
+def get_descriptions(request):
+    """Get the decriptions of the Solaris Porting Layer (SPL) or the
+    tunables, return with minimal formatting.
+    """
+
+    if request not in ('spl', 'zfs'):
+        print('ERROR: description of "{0}" requested)'.format(request))
+        sys.exit(1)
+
+    descs = {}
+    target_prefix = 'parm:'
+
+    command = ["/sbin/modinfo", request, "-0"]
+    info = subprocess.run(command, stdout=subprocess.PIPE,
+                          universal_newlines=True)
```

Comment:
On Centos 7.4, running python 3.4.5, I'm hitting the following error when running `arc_summary3.py` without arguments.  My guess is this  is caused by running zfs in-tree without installing kmods on the system. 

Can you update this function to read the module version from `/sys/module/{spl,zfs}/version` instead.  This will always work and is authoritative since it's the version of the module currently loaded in to the kernel.  That may differ from what `modinfo` reports if new kmods have been built but not yet loaded.

```
Traceback (most recent call last):
  File "./cmd/arc_summary/arc_summary3.py", line 833, in <module>
    main()
  File "./cmd/arc_summary/arc_summary3.py", line 790, in main
    print_header()
  File "./cmd/arc_summary/arc_summary3.py", line 354, in print_header
    zfs = get_version('zfs')
  File "./cmd/arc_summary/arc_summary3.py", line 327, in get_version
    modinfo = subprocess.run(command, stdout=subprocess.PIPE,
AttributeError: 'module' object has no attribute 'run'
```

Reply:
@behlendorf If this is my friend CentOS with Python 3.4, then the problem is that there it is using an old version of subprocess (seriously, what is it with CentOS?). See https://docs.python.org/3/library/subprocess.html , where it says "New in version 3.5". We could add a check at the beginning to make sure that it's 3.5 or later and fail gracefully; the good news is that this problem should go away by itself in a few months.

I assume we still want to check `/sys/module` and not `modinfo`? Will try to get to that over the weekend, but could be early next week. Thanks for the help!

Reply:
> I assume we still want to check /sys/module and not modinfo?

Yes, and it looks like `get_description` and `get_version` are the only callers ` subprocess` so perhaps we can avoid this compatibility issue entirely.

> the good news is that this problem should go away by itself in a few months.

Why is that?

Reply:
I'll see what I can do about `subprocess` next week ... I remember there was a work-around. -- I'm assuming that CentOS at some point will be upgrading Python 3 -- we're almost at Python 3.7, so 3.4 is pretty ancient.

Reply:
@scotws does look like CentOS provides both python 3.4 and 3.6 from the EPEL repository.  But if at all possible let's try and do something compatible with both.

Reply:
@behlendorf Sorry to bother you with this, but I can't seem to find a way to get the descriptions for the individual parameters from the `/sys/module/{spl,zfs}/` tree - only through `modinfo`. What am I missing? (Possibly this was the reason that the original `arc_summary.py` used `modinfo` for its version of `get_description`). Thanks!

Reply:
You're right, it looks like only the version is provided.  Well frankly these strings never change so while not ideal it wouldn't be terrible to hardcode a meaningful string in the utility.  The critical information is the version which is available.

Reply:
@behlendorf If we're going to hardcode the descriptions, we might as well make them better than they are now, with the real types (AFAIK, the bools are currently being displayed as ints). If possible, we could also include something along the lines of "less is better" or "should be 0". I'd just need somebody to proofread those very carefully who actually understands what is going on under the hood here ...

Reply:
Sorry, I badly misunderstood your comment before and I hope I didn't send you on a wild goose chase.  Initially I thought this simply refered to the module description itself, I see that's not the case you're talking about the parameter descriptions too.  You're right, that case isn't so simple and we don't want to hardcode that information since it'll be a significant maintence burden.  I'd be OK with any of the following solutions, with a preference for the first one.

* Update `get_description()` to work with python 3.4.

* Add `Requires: python >= 3.5` to spec file and somehow express in the script's shebang that at least version 3.5 needs to be used.  I'm not sure if that's possible since `/usr/bin/python3` doesn't seem to always link to the _newest_ python3 version, and it's possible to have `python34` and `python36` concurrently installed.

* Drop this functionality for now until we find the best way to handle it.

Reply:
As I understand it, the issue is that `subprocess.run()` doesn't exist on older versions of Python 3. The fix is trivial. Use the perfectly good older equivalents instead. Do this instead:
```
info = subprocess.check_output(command)
raw_output = info.decode(sys.stdin.encoding).split('\0')
```

Reply:
Full disclosure: `subprocess.run()` says it uses the `io.TextIOWrapper` default encoding, which says it uses `locale.getpreferredencoding(False)`. If you wanted to match more closely, you could use that instead of `sys.stdin.encoding` in my example. On my Ubuntu system, though, `sys.stdin.encoding` is `UTF-8` (from my terminal) and `locale.getpreferredencoding(False)` is `ANSI_X3.4-1968`. I have always used `sys.stdin.encoding` (or `stdout`) and it's always been fine for me. Using `sys.stdin.encoding` seems like the superior option.

Reply:
@rlaager Thanks for the suggestions! There are actually two issues at the moment: 

The first one is the `subprocess.run()` problem, which affects a very small number of machines -- if I'm seeing this correctly, it is only some CentOS versions. This will go away by itself as the last distributions catch up to Python 3.5, so I was actually planning to keep the current, high-level code and throw an `IF/THEN` version check to the effect of `< 3.5` around it with the older `subprocess.check_output()` for the stragglers as you suggest. When everybody is at 3.5 in a few months, we simply remove the `IF/THEN`. Yes, this is crude, but keeps the recommended, cleaner code for most machines, and is temporary anyway.

The second issue is that @behlendorf has pointed out we should be reading stuff from `/sys/module/{spl,zfs}/version` instead of using `modinfo` (like the original `arc_summary.py`) so we report what's actually in the kernel instead of what has been built but maybe not loaded (which I had honestly not thought of). That's also trivial, except that there seems no way to get the parameters descriptions from `/sys/module`. 

My current plan there is, if the user wants descriptions,  we load the parameters from `/sys/module` and use them as keys to a dictionary of parameters/descriptions we get from `modinfo`. Yes, that runs the risk that some descriptions might change between loaded/built-but-not-loaded, but that should be small. -- Unless I've missed something and there is some other way to retrieve the descriptions from `/sys/modules`, which would of course be best.

Thanks again!

Reply:
>  When everybody is at 3.5 in a few months.

I wouldn't get your hopes up for CentOS 7.  Since it's an enterprise distribution they're particuarly adverse to change and it wouldn't shock me if python 2.7 remains the default python version until it's [end-of-life in 2024](https://wiki.centos.org/About/Product).

The good news is that python 3.6 is also provided by the distribution, worst case we can do something distribution specific here and require `#!/bin/python36`.  But if we can fallback gracefully to something compatible that would be better.

Reply:
personally, I don't see check_output going away anytime soon, but...

there no need to check versions, check to see if you can import run from subprocess thusly:
```python
info = ''
if 'run' in dir(subprocess):
  info = subprocess.run(...)
elif 'check_output' in dir(subprocess):
  info = subprocess.check_output(...)
```

Reply:
@richardelling Thanks, hadn't thought of that, that's pretty much what I did now.